### ISSUE TITLE
sourcenet upkeep

### ISSUE THREAD

**Issue by jeff-regier** [prob-ml/bliss#24]
@ismael2395 @zzhaozhe-profolio  Here are a few idea for keeping `sourcenet.py` easy to read/maintain, based on what parts are missing unit test coverage. (None of these changes are particularly urgent.)

1.  It's generally better if train/eval mode are set outside of the class, within the train/eval loops (pytorch-lightning may expect that even), and within the unit tests. It can get confusing keeping track of which mode we're in if train/eval mode is set both within the class and outside of it.
https://github.com/applied-bayes/celeste/blob/9dc1ce03775e5e8bfe3c957fee96b4ecb571370a/celeste/models/sourcenet.py#L899-L901

1.  For that matter, it'd be better if sourcenet didn't have any awareness of whether it was in training or eval mode. Does the command below do something useful?
https://github.com/applied-bayes/celeste/blob/9dc1ce03775e5e8bfe3c957fee96b4ecb571370a/celeste/models/sourcenet.py#L814-L816
At prediction time, `with torch.no_grad()` should be called outside of `sourcenet.py` (in `train.py`), so that gradients aren't calculated anywhere (without having to call `detach`),

1.  Is `sample_class_weights` used anywhere? If not, can we delete it?
https://github.com/applied-bayes/celeste/blob/9dc1ce03775e5e8bfe3c957fee96b4ecb571370a/celeste/models/sourcenet.py#L35

1.  Can we just assert that `step` isn't none?
https://github.com/applied-bayes/celeste/blob/9dc1ce03775e5e8bfe3c957fee96b4ecb571370a/celeste/models/sourcenet.py#L62-L63

1.  Is the conditional below (and the ones that follow it in the code) for images that can't be exactly divided into tiles? If so, we should probably add a unit test for an image with an irregular size.
https://github.com/applied-bayes/celeste/blob/9dc1ce03775e5e8bfe3c957fee96b4ecb571370a/celeste/models/sourcenet.py#L66

1.  Are we forming a tensor below just to measure its size?
https://github.com/applied-bayes/celeste/blob/9dc1ce03775e5e8bfe3c957fee96b4ecb571370a/celeste/models/sourcenet.py#L396-L398

1.  Can we assert the following are always true?
https://github.com/applied-bayes/celeste/blob/9dc1ce03775e5e8bfe3c957fee96b4ecb571370a/celeste/models/sourcenet.py#L230
https://github.com/applied-bayes/celeste/blob/9dc1ce03775e5e8bfe3c957fee96b4ecb571370a/celeste/models/sourcenet.py#L820-L821
https://github.com/applied-bayes/celeste/blob/9dc1ce03775e5e8bfe3c957fee96b4ecb571370a/celeste/models/sourcenet.py#L850

1. Better to avoid calling `get_image_ptiles` if `locs` is `None` or if `source_params` is `None`. Then we won't need this ugly case:
https://github.com/applied-bayes/celeste/blob/9dc1ce03775e5e8bfe3c957fee96b4ecb571370a/celeste/models/sourcenet.py#L737-L749

1. Some of the function names is `sourcenet.py` begin with underscores and others don't, and there doesn't seem to be a pattern. For consistency, it might be easiest to start none of them with underscores. Is `forward` is the only "public" method in the class?

1. There are a number of `torch.long` tensors in sourcenet.py, which can be OK if they are storing indices. But GPUs excel at `torch.float32` arithmetic, and `torch.long` calculatation can be a bottleneck. Can some of these `torch.long` tensors be remove or converted to `torch.float32`?
